src/io/imzml/reader.rs

```diff
@@ -0,0 +1,1472 @@
+> {
+    reader: &'a mut ImzMLReaderType<R, S, C, D>,
+    index: usize,
+}
+
+impl<
+        'a,
+        R: Read + Seek,
+        S: Read + Seek,
+        C: CentroidLike + BuildFromArrayMap,
+        D: DeconvolutedCentroidLike + BuildFromArrayMap,
+    > ChromatogramIter<'a, R, S, C, D>
+{
+    pub fn new(reader: &'a mut ImzMLReaderType<R, S, C, D>) -> Self {
+        Self { reader, index: 0 }
+    }
+}
+
+impl<
+        R: Read + Seek,
+        S: Read + Seek,
+        C: CentroidLike + BuildFromArrayMap,
+        D: DeconvolutedCentroidLike + BuildFromArrayMap,
+    > Iterator for ChromatogramIter<'_, R, S, C, D>
+{
+    type Item = Chromatogram;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index < self.reader.chromatogram_index.len() {
+            let result = self.reader.get_chromatogram_by_index(self.index);
+            self.index += 1;
+            result
+        } else {
+            None
+        }
+    }
+}
+
+/// For open_file we assume that the IBD file is named the same as the imzML file
+impl<C: CentroidLike + BuildFromArrayMap, D: DeconvolutedCentroidLike + BuildFromArrayMap>
+    MZFileReader<C, D, MultiLayerSpectrum<C, D>> for ImzMLReaderType<fs::File, fs::File, C, D>
+{
+    fn open_file(_source: fs::File) -> io::Result<Self> {
+        #[cfg(feature = "filename")]
+        {
+            // Get the path from the file using the filename crate
+            let xml_path = filename::file_name(&_source)
+                .ok_or_else(|| {
+                    io::Error::new(io::ErrorKind::InvalidInput, "Source file has no path")
+                })?;
+                
+            // Derive IBD path from imzML path
+            let mut ibd_path = xml_path.with_extension("ibd");
+            if !ibd_path.exists() {
+                // Try .IBD (uppercase)
+                ibd_path = xml_path.with_extension("IBD");
+                if !ibd_path.exists() {
+                    return Err(io::Error::new(
+                        io::ErrorKind::NotFound,
+                        format!(
+                            "Could not find associated IBD file for imzML at {}",
+                            ibd_path.display()
+                        ),
+                    ));
+                }
+            }
+            
+            // Open the IBD file
+            let ibd_file = fs::File::open(&ibd_path)?;
+            
+            // Create the reader with both files
+            Ok(Self::new(_source, ibd_file))
+        }
+        #[cfg(not(feature = "filename"))]
+        {
+            // Cannot implement without filename feature
+            Err(io::Error::new(
+                io::ErrorKind::Unsupported,
+                "Cannot open imzML from File handle without the `filename` feature - use open_path() or ImzMLReaderType::new() directly"
+            ))
+        }
+    }
+
+    fn construct_index_from_stream(&mut self) -> u64 {
+        // Index should already be constructed during metadata parsing
```

Comment:
Neither `MzMLReaderType::build_index` nor `MzMLReaderType::construct_index_from_stream` appear to be called anywhere here.

Reply:
Fixed.